hw/xbox/nv2a.c

```diff
@@ -5012,36 +5015,89 @@ static void pgraph_method(NV2AState *d,
 
     case NV097_SET_VERTEX_DATA2F_M ...
             NV097_SET_VERTEX_DATA2F_M + 0x7c: {
-        slot = (class_method - NV097_SET_VERTEX_DATA2F_M) / 8;
-        unsigned int part =
-            ((class_method - NV097_SET_VERTEX_DATA2F_M) % 8) / 4;
-        pg->vertex_attributes[slot].inline_value[part] = *(float*)&parameter;
-        pg->vertex_attributes[slot].inline_value[2] = 0.0;
-        pg->vertex_attributes[slot].inline_value[3] = 1.0;
+        slot = (class_method - NV097_SET_VERTEX_DATA2F_M) / 4;
+        unsigned int part = slot % 2;
+        slot /= 2;
+        VertexAttribute *attribute = &pg->vertex_attributes[slot];
+        allocate_inline_buffer_vertices(pg, slot);
+        attribute->inline_value[part] = *(float*)&parameter;
+        /* FIXME: Should these really be set to 0.0 and 1.0 ? Conditions? */
+        attribute->inline_value[2] = 0.0;
+        attribute->inline_value[3] = 1.0;
+        if ((slot == 0) && (part == 1)) {
+            finish_inline_buffer_vertex(pg);
+        }
         break;
     }
-
     case NV097_SET_VERTEX_DATA4F_M ...
             NV097_SET_VERTEX_DATA4F_M + 0xfc: {
-        slot = (class_method - NV097_SET_VERTEX_DATA4F_M) / 16;
-        unsigned int part =
-            ((class_method - NV097_SET_VERTEX_DATA4F_M) % 16) / 4;
-        pg->vertex_attributes[slot].inline_value[part] = *(float*)&parameter;
+        slot = (class_method - NV097_SET_VERTEX_DATA4F_M) / 4;
+        unsigned int part = slot % 4;
+        slot /= 4;
+        VertexAttribute *attribute = &pg->vertex_attributes[slot];
+        allocate_inline_buffer_vertices(pg, slot);
+        attribute->inline_value[part] = *(float*)&parameter;
+        if ((slot == 0) && (part == 3)) {
+            finish_inline_buffer_vertex(pg);
+        }
+        break;
+    }
+    case NV097_SET_VERTEX_DATA2S ...
+            NV097_SET_VERTEX_DATA2S + 0x3c: {
+        slot = (class_method - NV097_SET_VERTEX_DATA2S) / 4;
+        assert(false); /* FIXME: Untested! */
+        VertexAttribute *attribute = &pg->vertex_attributes[slot];
+        allocate_inline_buffer_vertices(pg, slot);
+        /* FIXME: Is mapping to [-1,+1] correct? */
+        parameter ^= 0x80008000;
+        attribute->inline_value[0] = ((parameter & 0xFFFF)
```

Comment:
I'm not sure to what ranges shorts are mapped to. I'd expect [-1,+1] (inclusive).
Your code Would give wrong results for the max value 0x7FFF = 32767.0 which should probably map to 1.0 (if that's even what this setter is supposed to do). - Your code would map to 0.99996948...
My code handles that by mapping [-32768,32767] to [0,65535], then to [0,1], [-0.5,0.5] and then finally scaling to [-1,1]. (ignoring all float inaccuracy for now).